Lua: add vim.validate() to check function parameters

h-michael opened the pull request:
related discussions:
https://github.com/neovim/neovim/pull/11248#pullrequestreview-304232037
https://github.com/neovim/neovim/pull/11276#discussion_r339368423

Depending on the purpose of use, it may be excessive or insufficient, so I wanted to solicit opinions and created a PR once.

h-michael committed:
Lua: vim.validate()

We often want to do type checking of public function arguments.

- test: Rename utility_function_spec.lua to vim_spec.lua
- .luacov: Map lua module names

h-michael commented:
umm, I think we get wrong code coverage result.

justinmk commented on test/functional/lua/vim_spec.lua:
```diff
@@ -327,4 +327,61 @@ describe('lua stdlib', function()
   end)
+
+  it('vim.validate', function()
+    eq(true, pcall(exec_lua, "return vim.validate({ arg1={ {}, 'table' }})"))
```
There's no reason to use `pcall` if you expect success. `pcall` is only for testing an expected failure.

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -190,4 +190,98 @@ function vim.pesc(s)
+---     - thread
+---     - userdata
+function vim.validate(...)
+  local function _type_name(t)
```
this function seems unnecessary. Maybe if the type aliases were 1 char I guess ('t' for table, 's', for string, ...?). But even then, it doesn't really improve readability or save much typing.

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -190,4 +190,98 @@ function vim.pesc(s)
+  for _, arg in pairs(args) do
+    for arg_name, v in pairs(arg) do
+      assert(type(arg_name) == 'string',string.format('Expected string, got %s', type(arg_name)))
+      assert(type(v) == 'table', string.format('Expected table, got %s', type(v)))
```
why not use `vim.validate` ? :)

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -190,4 +190,98 @@ function vim.pesc(s)
+      local is_optional = v[3] == true
+
+      if is_optional then
+        assert(_check_type(v[1], expected_type) or actual_arg_type == 'nil', string.format("'%s' is expected %s or nil, got %s", arg_name, expected_type, actual_arg_type))
```
don't need a separate message for optional types. If they're optional they are simply ignored if missing, otherwise they are type-checked.

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -190,4 +190,98 @@ function vim.pesc(s)
+      if is_optional then
+        assert(_check_type(v[1], expected_type) or actual_arg_type == 'nil', string.format("'%s' is expected %s or nil, got %s", arg_name, expected_type, actual_arg_type))
+      else
+        assert(_check_type(v[1], expected_type), string.format("'%s' is expected %s, got %s", arg_name, expected_type, actual_arg_type))
```
```suggestion
        assert(_check_type(v[1], expected_type), string.format("%s: expected %s, got %s", arg_name, expected_type, actual_arg_type))
```

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -190,4 +190,98 @@ function vim.pesc(s)
+--- </pre>
+---
+---@param ... Table or list of table. That table is "argument_name = { validation_target, type_name, expected_description }".
+function vim.validate_with_func(...)
```
```suggestion
function vim.validate_with_func(opt)
  assert(type(opt) == 'table')
```

justinmk replied:
Actually, why is `validate_with_func` a separate function? `vim.validate()` could accept a function instead of a boolean for the "optional" field.

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -190,4 +190,98 @@ function vim.pesc(s)
+--- </pre>
+---
+---@param ... Table or list of table. That table is "argument_name = { validation_target, type_name, expected_description }".
+function vim.validate_with_func(...)
```
why is `validate_with_func` a separate function? `vim.validate()` could accept a function instead of a boolean for the "optional" field.

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -190,4 +190,98 @@ function vim.pesc(s)
+---     - nil
+---     - thread
+---     - userdata
+function vim.validate(...)
```
I don't see a reason for this to be variadic (`...`). Seems like a table can/should be required.
```suggestion
function vim.validate(opt)
  assert(type(opt) == 'table')
```

h-michael commented:
I've fixed.

h-michael commented:
I've also fixed .luacov. https://github.com/neovim/neovim/pull/11307/commits/eb9f4ddd38407a1fb421c0f0b2539cc11171d546

[before]
https://codecov.io/gh/neovim/neovim/compare/5a212a96a6d784c9d62eee2bbca525ff0f815061...d213223af07a49b0f3453c09e19488dca05f2428/src/runtime/lua/vim/shared.lua?before=runtime/lua/vim/shared.lua

[after]
https://codecov.io/gh/neovim/neovim/compare/75ad5587477b1fc90cb0c693794ee7053c3339c7...eb9f4ddd38407a1fb421c0f0b2539cc11171d546/src/runtime/lua/vim/shared.lua?before=runtime/lua/vim/shared.lua

cc: @blueyed

h-michael commented:
@justinmk Is there any other problem to merge this?

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -190,4 +190,78 @@ function vim.pesc(s)
+---  validate({ arg={ { 'foo' }, 'table' }})                                     --> Nop
+---  validate({ arg={ 1, 'table' } })                                            --> error("arg: expected table, got number")
+---  validate({ arg1={ { 'foo' }, 'table' }, arg2={ 1, 'string' } })             --> error("arg2: expected string, got number")
+---  validate({ arg={ 3, function(a) return (a % 2) == 0  end, 'even number' }}) --> error("arg: expected even number, got 3")
```
here is says we always call error(), but actually we do assert() in some cases. Maybe it's useful to always do error().

I will do this when merging.

h-michael replied:
I've fixed it :)

justinmk committed:
Lua: vim.validate()

justinmk committed:
Lua: Use vim.validate() instead of assert()

justinmk committed:
fix nvim__buf_stats

Accidentally renamed in 1cb467454782.

justinmk commented:
pushed an update with these changes:

- docs, naming
- use `vim.validate()` in some places where we were using `assert()` (nice code reduction!)
- `_is_type`: use a map of `type_names` instead of if-then chain
- simplify `vim.validate()` impl

justinmk committed:
Lua: mark some functions as "private"

Problem:  scripts/gen_vimdoc.py gets confused and tries to generate docs
          for `fn_index` and `func`.
Solution: Rename them to be private.

justinmk commented on src/nvim/lua/vim.lua:
```diff
@@ -261,14 +261,13 @@ end
+  return t[key]
 end
-local fn = setmetatable({}, {__index=fn_index})
+local fn = setmetatable({}, {__index=_fn_index})
```
@bfredl JFI: renaming/rearranging these so that `gen_vimdoc.py` does not think they're "public" (and it makes semantic sense anyway)

h-michael commented:
@justinmk Thank you for follow up :)
